### PR TITLE
Introduce Map#_requestRenderFrame

### LINKED ISSUE #6233: Refactor handlers to use a `requestRenderFrame` callback rather than `Camera#_startAnimation`
#6005 introduced `Camera#_startAnimation` as a hook to allow interaction handlers to perform map updates on a schedule controlled by the render loop (rather than having their updates be interleaved with render frames).  This approach is problematic because it assumes that only a single interaction handler or programmatic animation "owns" the camera for the duration of any given render frame.  As @jfirebaugh points out, this makes handler behavior sensitive to event timing:
> For instance, if you receive both drag and scroll events within a single render tick, or receive mousedown, mousemove, and mouseup within a single render tick, things go awry

A better design would be to expose `Map#_requestRenderFrame(cb)`, allowing handlers or animations to register a callback to be called during the next render frame.  In addition to fixing the issues described above, this would allow the camera and handler code to be written just like it would be if we were using `requestAnimationFrame()`, which will make it clearer and more readable.

### PR DESCRIPTION
**Pull request by anandthakker**
Fixes #6233

- [x] Update "can interleave with another handler" unit tests (`map.stop()` no longer correctly simulates another handler taking over)
- [ ] Add tests confirming that handlers are indifferent to relative timing of mouse events & render frames (or mouse events for different handlers)

### CODE REVIEW

File: src/ui/handler/drag_pan.js [hunk 1]
```diff
@@ -143,6 +144,19 @@ class DragPanHandler {
...
+            this._frameId = this._map._requestRenderFrame(this._onDragFrame);
+        }
+    }
+
+    _cancelFrame() {
+        if (this._frameId) {
+            this._map._cancelRenderFrame(this._frameId);
+            this._frameId = null;
+        }
+    }
```
Comment (anandthakker) [inline src/ui/handler/drag_pan.js]: TODO: inline `_requestFrame` and `_cancelFrame`

**Comment by anandthakker** [comment]
> Add tests confirming that handlers are indifferent to relative timing of mouse events & render frames

This is not 100% true at the moment: if we get a `mousedown`, `mousemove` and `mouseup` all in the same frame, the current implementation in this branch would not perform a `drag`, because it still relies on doing so in a render frame.  So, in this case, we'd have `dragstart`, `dragend` (no `drag`), whereas if the render frame occurred before the `mouseup`, we'd have `dragstart`, `drag`, `dragend`

File: test/unit/util/task_queue.test.js [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        const q = new TaskQueue();
+        const cb = t.spy();
+        const id = q.add(cb);
+        q.run();
+        q.remove(id);
+        t.equal(cb.callCount, 1);
+        t.end();
+    });
+
+    t.test('Does not add tasks to the currently-running queue', (t) => {
```
[the remark this answers is not in the document]
Comment (anandthakker) [reply test/unit/util/task_queue.test.js]: Good catch

**Comment by anandthakker** [comment]
@jfirebaugh just pushed a change that addresses https://github.com/mapbox/mapbox-gl-js/pull/6328#issuecomment-372878446 -- re-review?  If this looks alright, I'll propagate the approach to `TouchZoomRotate` as well

**Comment by jfirebaugh** [comment]
I think I'd prefer an approach where `onMouseUp` _always_ schedules the inertial animation via `_requestRenderFrame`, so that there's a single execution path rather than two. (Trying really hard to keep the complexity as low as possible, since that's still relatively high.)

**Comment by anandthakker** [comment]
@jfirebaugh good call -- that does sound simpler...  but I'm realizing that there's a slightly different timing issue that's still not addressed here: `mouseup, mousedown` in the same render tick (i.e. ending one drag and starting a new one) or, even a bit worse, `mousemove, mouseup, mousedown` in the same tick.   It seems like dealing with these case "correctly" would require also _starting_ the drag in a `_requestRenderFrame` callback rather than synchronously in `onMouseDown`... which means further complicating state management.   

I'm inclined to just punt on this issue: revert to always ending a drag (and kicking off the inertial animation) immediately in `_onMouseUp` -- if a `mousemove` snuck into the same render frame behind the `mouseup`, we just cancel the scheduled drag frame.  It looks like this is what Leaflet does,  too: https://github.com/Leaflet/Leaflet/blob/64712282602be02535ebee9ebf5cb64f789898d4/src/dom/Draggable.js#L213-L220 (cc @mourner )

**Comment by jfirebaugh** [comment]
So just revert the last commit here? SGTM.

File: test/unit/util/task_queue.test.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (anandthakker) [reply test/unit/util/task_queue.test.js]: Fixed.

**Pull request merged**